plugins/calibrecompanion.koplugin/sha1.lua

```diff
@@ -0,0 +1,330 @@
+-- $Revision: 1.5 $
+-- $Date: 2014-09-10 16:54:25 $
+ 
+-- This module was originally taken from http://cube3d.de/uploads/Main/sha1.txt.
+ 
+-------------------------------------------------------------------------------
+-- SHA-1 secure hash computation, and HMAC-SHA1 signature computation,
+-- in pure Lua (tested on Lua 5.1)
+-- License: MIT
+--
+-- Usage:
+-- local hashAsHex = sha1.hex(message) -- returns a hex string
+-- local hashAsData = sha1.bin(message) -- returns raw bytes
+--
+-- local hmacAsHex = sha1.hmacHex(key, message) -- hex string
+-- local hmacAsData = sha1.hmacBin(key, message) -- raw bytes
+--
+--
+-- Pass sha1.hex() a string, and it returns a hash as a 40-character hex string.
+-- For example, the call
+--
+-- local hash = sha1.hex("iNTERFACEWARE")
+--
+-- puts the 40-character string
+--
+-- "e76705ffb88a291a0d2f9710a5471936791b4819"
+--
+-- into the variable 'hash'
+--
+-- Pass sha1.hmacHex() a key and a message, and it returns the signature as a
+-- 40-byte hex string.
+--
+--
+-- The two "bin" versions do the same, but return the 20-byte string of raw
+-- data that the 40-byte hex strings represent.
+--
+-------------------------------------------------------------------------------
+--
+-- Description
+-- Due to the lack of bitwise operations in 5.1, this version uses numbers to
```

Comment:
Sidebar: That applies to Lua 5.1, not LuaJIT, which ships a JIT-enabled bitop module by default ;).

Reply:
I skimmed through it really quickly and I failed to notice it wasn't using bit.bor.

On a related note: https://github.com/Egor-Skriptunoff/pure_lua_SHA/blob/304d4121f080e68ef209d3f5fe093e5a955a4978/sha2.lua

> this is the fastest pure Lua implementation of SHA you can find.

It also does MD5. This might be worth checking out. (MIT license)

Reply:
Ooh, shiny :). And it does partial MD5, which we need.